src/qobj/arithmetic_and_attributes.jl

```diff
@@ -514,18 +514,38 @@ Quantum Object:   type=Operator   dims=[2]   size=(2, 2)   ishermitian=true
 ```
 """
 function ptrace(QO::QuantumObject{<:AbstractArray,KetQuantumObject}, sel::Union{AbstractVector{Int},Tuple})
-    length(QO.dims) == 1 && return QO
+    ns = length(sel)
+    if ns == 0
+        return tr(QO * QO')
+    else
+        nd = length(QO.dims)
+        !all(nd .>= sel .> 0) && throw(
+            ArgumentError("Invalid indices in `sel`: $(sel), the given QuantumObject only have $(nd) sub-systems"),
+        )
+        ns != length(unique(sel)) && throw(ArgumentError("Duplicate selection indices in `sel`: $(sel)"))
+        nd == 1 && return QO * QO'  # ptrace should always return Operator
```

Comment:
Also, here. Why are you returning `QO * QO’`? Is it a qutip functionality too?

Reply:
Ok, I see that you return the density matrix. Why don’t you use `ket2dm` so?

Reply:
okay, I changed to `ket2dm`